### DISCUSSION TITLE
In-memory management and state sync

### DISCUSSION THREAD

**Discussion by hyperserac** [ava-labs/hypersdk#932] [Chat]
Any logic that sits inside the Controller can modify the memory of a node (tokenvm example)
As far as i understand the transaction lifecycle, the logic inside the Controller is executed only after a transaction is marked as `Accepted` : 

- how can we make sure that the expected behaviour is actually executed in the Controller ? 
- what if there are logic issues in the Controller ?
I think any in-memory changes should be persisted on the state, such that all the nodes on the network can sync their memory after a block has been produced. 
Maybe the commands (inputs) and the events (outputs) logging could be managed by the sdk so that we have a standard on state synchronisation

**Comment by patrick-ogrady** [comment]
> how can we make sure that the expected behaviour is actually executed in the Controller ?

I've been thinking a lot about this flow lately. I think the most robust option (to avoid potential misuse) would be to remove this capability and require such mechanisms to be implemented over the reliable stream exported by the VM (https://github.com/ava-labs/hypersdk/issues/961).

> what if there are logic issues in the Controller ?
I think any in-memory changes should be persisted on the state, such that all the nodes on the network can sync their memory after a block has been produced.
Maybe the commands (inputs) and the events (outputs) logging could be managed by the sdk so that we have a standard on state synchronisation

Anything that is required for state transition computation **must** be stored in state. You should not be using anything updated in `Controller` in an `Action` (it will be non-deterministic because it is updated as consensus runs).

I've created an issue that should make this easier to reason about (will update `tokenvm` with a clearer example as well).
